### PR TITLE
Enable warm starting of the model for forecast-only mode

### LINKED ISSUE #2016: GEFS Warm Start Capability
### What new functionality do you need?

Add ability to start GEFS free forecasts using warmstarts 

### What are the requirements for the new functionality?

- additional warm restarts
- edits for configuration 

### Acceptance Criteria

successful GEFS free forecast using warm start ICs using S2S model version 

### Suggest a solution (optional)

_No response_

### PR DESCRIPTION
**Pull request by aerorahul**
# Description

This PR:
- adds an option to `warm|cold` start the ufs-weather-model when setting up an experiment for GFS/GEFS
- adds the capability to stage FV3 restarts when the atmosphere is warm started from FV3 restarts.  This would be used in the GEFS free-forecast for part of the reforecast capability.
- If a coupled model is warm started for the atmosphere, also check if mediator restarts are present.  If so, stage mediator restarts to allow restarting the coupled model.

A couple of `TODO` are introduced:
- are mediator restarts only available in the `APP=S2S+` mode?  What about `ATMW`?

In a future PR, if desired a test can be added.
Will coordinate testing from mock atmosphere restarts for the ensemble with @NeilBarton-NOAA 

Fixes #2016 

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? YES [new capability should be documented]

# How has this been tested?
Free forecast from cold-start ICs is not broken.  This will be tested in the CI.
Warm start capability testing will be coordinated w/ @NeilBarton-NOAA 

# Checklist
- [ ] Any dependent changes have been merged and published
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] I have made corresponding changes to the documentation if necessary

### CODE REVIEW

**Comment by aerorahul** [comment]
@WalterKolczynski-NOAA 
Can we mock C48mx500 ensemble warm starts for testing this branch?  A single machine would be fine.  I followed the same  structure as ocean restarts.

**Comment by WalterKolczynski-NOAA** [comment]
@aerorahul Try now on Hera. If that works I'll mirror.

**Comment by NeilBarton-NOAA** [comment]
@WalterKolczynski-NOAA I don't see the staged ICs on hera? Did you transfer them? if so, could you point me to them? Thanks

**Comment by aerorahul** [comment]
@NeilBarton-NOAA 
I was able to mock C48 and 5-deg restarts from one of the SOCA provided ICs and try them in an experiment for staging.
The `S2S` experiment `gefs_warm` on Hera is at `/scratch1/NCEPDEV/stmp2/Rahul.Mahajan/RUNTESTS/EXPDIR/gefs_warm` and the mock ICs at `/scratch1/NCEPDEV/stmp2/Rahul.Mahajan/RUNTESTS/EXPDIR/gefs_warm/ICs/gefs_test`.

The staging works as expected.  See the experiment `COM` at `/scratch1/NCEPDEV/stmp2/Rahul.Mahajan/RUNTESTS/COMROT/gefs_warm/gefs.20210323/`

However, the model failed and I am not familiar with the (GEFS) configuration.
A model forecast log file for `mem000` is at `/scratch1/NCEPDEV/stmp2/Rahul.Mahajan/RUNTESTS/COMROT/gefs_warm/logs/2021032312/gefsfcst.log`
The corresponding model run directory is at:
`/scratch1/NCEPDEV/stmp2/Rahul.Mahajan/RUNDIRS/gefs_warm/fcst.280268`

If someone could take a look, I would appreciate the assistance.

The objective of this PR is to test staging and I think that works as expected.

**Comment by NeilBarton-NOAA** [comment]
@aerorahul and/or @WalterKolczynski-NOAA 

The warm start files at 
hera:/scratch1/NCEPDEV/global/glopara/data/ICSDIR/prototype_ICs/gefs_test/2021032312/mem00[0,1,2]/atmos
are broken links for me. 

Could you double check these?

**Comment by WalterKolczynski-NOAA** [comment]
@NeilBarton-NOAA Should be fixed now. Don't know why the links broke.

**Comment by NeilBarton-NOAA** [comment]
For this to run out-of-the-box, 
```
--- a/parm/config/gefs/config.base.emc.dyn
+++ b/parm/config/gefs/config.base.emc.dyn
@@ -286,7 +286,7 @@ export MEMDIR="mem${ENSMEM}"
 if [[ "${EXP_WARM_START}" = ".false." ]]; then
     export IAU_FHROT=0
 else
-    export IAU_FHROT=3
+    export IAU_FHROT=0
 fi
```

**Comment by aerorahul** [comment]
> For this to run out-of-the-box,
> 
> --- a/parm/config/gefs/config.base.emc.dyn +++ b/parm/config/gefs/config.base.emc.dyn @@ -286,7 +286,7 @@ export MEMDIR="mem${ENSMEM}" if [[ "${EXP_WARM_START}" = ".false." ]]; then export IAU_FHROT=0 else
> 
> * export IAU_FHROT=3
> 
> * export IAU_FHROT=0
>   fi

@NeilBarton-NOAA 
I will open an issue for this and work on it.

**Review by WalterKolczynski-NOAA — APPROVED** [review]

**Review by NeilBarton-NOAA — APPROVED** [review]

**Comment by aerorahul** [comment]
> Experiment C48_S2SA_gefs_c0d79096 **SUCCESS** Tue Nov 14 18:21:17 UTC 2023

@WalterKolczynski-NOAA This is the only test that is impacted by this PR.  Save some cycles perhaps?

**Comment by aerorahul** [comment]
> For this to run out-of-the-box,
> 
> --- a/parm/config/gefs/config.base.emc.dyn +++ b/parm/config/gefs/config.base.emc.dyn @@ -286,7 +286,7 @@ export MEMDIR="mem${ENSMEM}" if [[ "${EXP_WARM_START}" = ".false." ]]; then export IAU_FHROT=0 else
> 
> * export IAU_FHROT=3
> 
> * export IAU_FHROT=0
>   fi

@NeilBarton-NOAA 
Can you post the diff again?  I think the markdown has modified the message here.

**Comment by aerorahul** [comment]
> > For this to run out-of-the-box,
> > --- a/parm/config/gefs/config.base.emc.dyn +++ b/parm/config/gefs/config.base.emc.dyn @@ -286,7 +286,7 @@ export MEMDIR="mem${ENSMEM}" if [[ "${EXP_WARM_START}" = ".false." ]]; then export IAU_FHROT=0 else
> > 
> > * export IAU_FHROT=3
> > * export IAU_FHROT=0
> >   fi
> 
> @NeilBarton-NOAA Can you post the diff again? I think the markdown has modified the message here.

Nevermind, I edited your comment to include a code-block.

**Comment by aerorahul** [comment]
> ```
> Experiment C96_atm3DVar_c0d79096 Terminated: *** FAILED ***
> Experiment C96_atm3DVar_c0d79096 Terminated with 1 tasks failed at Tue Nov 14 12:26:17 CST 2023
> Error logs:
> /work2/noaa/stmp/GFS_CI_ROOT/PR/2031/RUNTESTS/COMROT/C96_atm3DVar_c0d79096/logs/2021122100/gfsanal.log
> ```

@WalterKolczynski-NOAA 
Not sure why this failed.  This is not related to the changes in this PR.

**Comment by WalterKolczynski-NOAA** [comment]
I'm looking now.

**Comment by WalterKolczynski-NOAA** [comment]
Time limit. Has this branch pulled in the changes that slurm fix for analysis on Orion?

**Comment by aerorahul** [comment]
> Time limit. Has this branch pulled in the changes that slurm fix for analysis on Orion?

probably not.
I will update.

**Comment by aerorahul** [comment]
@NeilBarton-NOAA I included your fix for running out of the box in this PR itself.  Thanks for finding and making the suggestion.

File: scripts/exglobal_stage_ic.sh [hunk 1]
```diff
@@ -66,7 +89,7 @@ for MEMDIR in "${MEMDIR_ARRAY[@]}"; do
         ;;
       "025" )
         for nn in $(seq 1 3); do
-          src="${BASE_CPLIC}/${CPL_OCNIC}/${PDY}${cyc}/${MEMDIR}/ocean/${PDY}.${cyc}0000.MOM.res_${nn}.nc"
+          src="${BASE_CPLIC}/${CPL_OCNIC:-}/${PDY}${cyc}/${MEMDIR}/ocean/${PDY}.${cyc}0000.MOM.res_${nn}.nc"
           tgt="${COM_OCEAN_RESTART}/${PDY}.${cyc}0000.MOM.res_${nn}.nc"
```
Comment (JessicaMeixner-NOAA) [inline scripts/exglobal_stage_ic.sh:93]: @aerorahul should this be COM_OCEAN_RESTART_PREV?   I'm getting an error: 
`scripts/exglobal_stage_ic.sh: line 93: COM_OCEAN_RESTART: unbound variable`

Comment (aerorahul) [reply scripts/exglobal_stage_ic.sh:93]: It absolutely should be.  
Apologies for missing this (and the low res test also missed this). 
I will issue a hotfix asap.

Comment (JessicaMeixner-NOAA) [reply scripts/exglobal_stage_ic.sh:93]: Thanks for making the hotfix!  I can confirm that this change worked.  The forecast wasn't successful, but that was for other reasons - not this :)

Comment (aerorahul) [reply scripts/exglobal_stage_ic.sh:93]: Fix is in #2070

**Pull request merged**